### PR TITLE
fix(amf): PDU session establishment reject with cause DNN not supported

### LINKED ISSUE #11429: [5G AGW] [Spirent] [Scale test] -DNN not supported issue is triggering while executing scale Test with 250K UDP downlink traffic per UE
-Version:1.7.0-1643771447-32938190**

While performing scale test in Spirent, with 250K UDP Downlink traffic per UE for some IMSI observing DNN not supported, even after the DNN configured same as in NMS board. For this reason we are observing PDU Establishment request failures. Please check the logs attached below.  

(https://app.zenhub.com/files/170803235/d4f5d4b0-af11-4167-83e1-00ac5d9218bc/download)[https://app.zenhub.com/files/170803235/c469421b-85ef-405c-aee8-84d5c40838bc/download]

(https://app.zenhub.com/files/170803235/c469421b-85ef-405c-aee8-84d5c40838bc/download)[https://app.zenhub.com/files/170803235/7ff544e8-bc4f-4a5f-8938-40a0f911a4d6/download]

(https://app.zenhub.com/files/170803235/7ff544e8-bc4f-4a5f-8938-40a0f911a4d6/download)rcontent.com/61bb050c5a4b60928b83af15/3ec93e1b-8593-4bf3-9cd5-ddc57eb43322)

### PR DESCRIPTION
**Pull request by Chandhu-Wavelabs**
Signed-off-by: chandhu-wavelabs <chandhu.naga@wavelabs.ai>

## Summary
In few of the scale tests, s6a response (having APN and AMBR) is send by subscriberdb after getting the PDU session
establishment request which was resulting in session failure.

Code changes are done to initiate the request early on to ensure AMF has all the required information to process
PDU Session creation
 ## Test Plan
 Scale test is passed with this fix. Attached one random UE - AGW message flow from scale test results.
 
image

UT cases are passed.
image
image

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by panyogesh** [comment]
Can you please change the heading : like 
"PDU session establishment reject with cause DNN not supported"

Also please put the fix summary on changes done

File: lte/gateway/python/setup.py [hunk 1]
```diff
@@ -88,6 +88,8 @@
         'scripts/dp_probe_cli.py',
         'scripts/user_trace_cli.py',
         'scripts/icmpv6.py',
+        'scripts/test_supi_profile_cli.py',
```
Comment (ssanadhya) [inline lte/gateway/python/setup.py]: Please separate these changes into a different PR.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp [hunk 2]
```diff
@@ -839,6 +844,10 @@ TEST_F(AMFAppStatelessTest, TestAfterPDUSessionEstReq) {
                                                sizeof(ue_smc_response_hexbuf));
   EXPECT_EQ(rc, RETURNok);
 
+  s6a_update_location_ans_t ula_ans = util_amf_send_s6a_ula(imsi);
+  rc = amf_handle_s6a_update_location_ans(&ula_ans);
+  EXPECT_TRUE(rc == RETURNok);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  EXPECT_EQ(rc, RETURNok);
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp [hunk 3]
```diff
@@ -708,6 +709,10 @@ TEST_F(AMFAppStatelessTest, TestAfterRegistrationComplete) {
...
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Use Expect_EQ wherever possible so that you can get an informative error message in case of failure.
```suggestion
  EXPECT_EQ(rc, RETURNok);
```

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 4]
```diff
@@ -1651,14 +1709,18 @@ TEST_F(AMFAppProcedureTest, ServiceRequestSignalling) {
                                                sizeof(ue_smc_response_hexbuf));
   EXPECT_TRUE(rc == RETURNok);
 
+  send_initial_context_response(amf_app_desc_p, ue_id);
+
+  s6a_update_location_ans_t ula_ans = util_amf_send_s6a_ula(imsi);
+  rc = amf_handle_s6a_update_location_ans(&ula_ans);
+  EXPECT_TRUE(rc == RETURNok);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp]: Use Expect_EQ wherever possible so that you can get an informative error message in case of failure.
```suggestion
  EXPECT_EQ(rc, RETURNok);
```

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/python/setup.py (hunk 1 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/python/setup.py]: raised new PR #12248

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp (hunk 2 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp (hunk 4 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp]: addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp (hunk 3 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp [hunk 5]
```diff
@@ -90,10 +90,11 @@ int amf_handle_security_complete_response(
       }
     }
 
-    /* FSM takes care of sending initial context setup request */
-    ue_state_handle_message_initial(COMMON_PROCEDURE_INITIATED1,
-                                    STATE_EVENT_SEC_MODE_COMPLETE, SESSION_NULL,
-                                    ue_mm_context, amf_ctx);
+    // s6a update location request
+    int rc = amf_send_n11_update_location_req(ue_mm_context->amf_ue_ngap_id);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: What does "n11" indicates in this function name, amf_send_n11_update_location_req. n11 is the interface between AMF and SMF. Correct me if I am wrong

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp [hunk 6]
```diff
@@ -90,10 +90,11 @@ int amf_handle_security_complete_response(
...
+    if (rc == RETURNerror) {
+      OAILOG_INFO(LOG_AMF_APP, "AMF_APP: n11_update_location_req failure\n");
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: Change the log level to ERROR and message description shall indicate that "AMF has failed to send Update Location Request ".
While logging include amf_ue_ngap_id and imsi64 for better debugging.

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp [hunk 7]
```diff
@@ -90,10 +90,11 @@ int amf_handle_security_complete_response(
       }
     }
 
-    /* FSM takes care of sending initial context setup request */
-    ue_state_handle_message_initial(COMMON_PROCEDURE_INITIATED1,
-                                    STATE_EVENT_SEC_MODE_COMPLETE, SESSION_NULL,
-                                    ue_mm_context, amf_ctx);
+    // s6a update location request
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: Comment can be changed to "Send s6a update location request"

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp (hunk 5 above)
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: Why rc is re-declared? As it is already initialized at the beginning of this fun(). 
amf_send_n11_update_location_req(), can be directly called within if condition and check for return value

File: lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp [hunk 8]
```diff
@@ -645,6 +644,9 @@ int amf_decrypt_imsi_info_answer(itti_amf_decrypted_imsi_info_ans_t* aia) {
                                (supi_imsi.plmn.mnc_digit3 & 0xf);
   }
 
+  imsi64 = amf_imsi_to_imsi64(params->imsi);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp:886]: Since imsi64 is obtained here, it can be removed from line number 656.

Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp:886]: At line number, https://github.com/magma/magma/pull/12088/files#diff-359e73baac363f6d67312a481dde16bd3d45d72538d19589bb94f57e5ac29385R593 imsi64 is already fetched; 
Can we update that value to ue-context?

File: lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp [hunk 9]
```diff
@@ -680,8 +682,8 @@ int amf_decrypt_imsi_info_answer(itti_amf_decrypted_imsi_info_ans_t* aia) {
    * Execute the requested new UE registration procedure
    * This will initiate identity req in DL.
    */
-  rc = amf_proc_registration_request(aia->ue_id, is_amf_ctx_new, params);
-  OAILOG_FUNC_RETURN(LOG_NAS_AMF, rc);
+  amf_proc_registration_request(aia->ue_id, is_amf_ctx_new, params);
+  return imsi64;
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp]: Replace the return statement with, OAILOG_FUNC_RETURN and return the imsi64

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp (hunk 5 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: amf_send_n11_update_location_req this function name is carry forwarded from existing code, need to re-structure the code related to it and we will raise a separate PR.

Yes, N11 is interface between AMF and SMF.

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp (hunk 6 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp (hunk 7 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp (hunk 5 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_Security_Mode.cpp]: addressed

File: lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp (hunk 8 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp:886]: addressed

File: lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp (hunk 9 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp]: addressed

File: lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp [hunk 10]
```diff
@@ -883,21 +882,23 @@ int amf_decrypt_imsi_info_answer(itti_amf_decrypted_imsi_info_ans_t* aia) {
...
 
   params->decode_status = ue_context->amf_context.decode_status;
   /*
    * Execute the requested new UE registration procedure
    * This will initiate identity req in DL.
    */
-  rc = amf_proc_registration_request(aia->ue_id, is_amf_ctx_new, params);
-  OAILOG_FUNC_RETURN(LOG_NAS_AMF, rc);
+  amf_proc_registration_request(aia->ue_id, is_amf_ctx_new, params);
+  OAILOG_FUNC_RETURN(LOG_NAS_AMF, imsi64);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp]: Could change the return value at line, https://github.com/magma/magma/pull/12088/files#diff-359e73baac363f6d67312a481dde16bd3d45d72538d19589bb94f57e5ac29385R842

Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp]: Module needs to be same for this function. Entry and return statements are using two different module names

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 11]
```diff
@@ -437,6 +436,10 @@ TEST_F(AMFAppProcedureTest, TestRegistrationProcNoTMSI) {
                                                sizeof(ue_smc_response_hexbuf));
   EXPECT_TRUE(rc == RETURNok);
 
+  s6a_update_location_ans_t ula_ans = util_amf_send_s6a_ula(imsi);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:440]: Can we add more scenario, like currently it's assumed that HSS always sends DIAMETER_SUCCESS. 
Scenario-1: If result code other than DIAMETER_SUCCESS
Scenario-2: If HSS didn't provide apn
Scenario-2: If imsi doesn't match

File: lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp (hunk 10 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp]: addressed

Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp]: addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp (hunk 11 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:440]: this test case only for checking the TestRegistrationProcNoTMSI use case. For the negative scenarios which you have mentioned are already taken care in other UT cases.

You can find them in amf_app_test_util.cpp:send_proc_authentication_info_answer().

**Comment by ssanadhya** [comment]
@rsarwad , can you steer this PR for review? 

@Chandhu-Wavelabs , I do not have any new comments here. Let's wait on @rsarwad to review.

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp (hunk 11 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:440]: As part of amf_app_test_util.cpp:send_proc_authentication_info_answer(), daimeter cause is taken care for s6a-Authentication information procedure. My concern is, can we add similar case for Update location procedure. Because resule code is not taken care at, https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/amf/nas_proc.cpp#L903
On reception of Update Location Ans, amf shall check for result code, if it's not success. I think amf should reject the Registration request

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 12]
```diff
@@ -2472,4 +2554,85 @@ TEST_F(AMFAppProcedureTest, TestAuthFailureFromSubscribeDbLock) {
...
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for security mode complete response from UE */
+  rc = send_uplink_nas_message_ue_smc_response(amf_app_desc_p, ue_id, plmn,
+                                               ue_smc_response_hexbuf,
+                                               sizeof(ue_smc_response_hexbuf));
+  EXPECT_TRUE(rc == RETURNok);
+
+  s6a_update_location_ans_t ula_ans = util_amf_send_s6a_ula(imsi);
+  strcpy(ula_ans.imsi, "123456789012345");
```
[the remark this answers is not in the document]
Comment (rsarwad) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp]: Address the cpplint warning messages

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 13]
```diff
@@ -2472,4 +2554,85 @@ TEST_F(AMFAppProcedureTest, TestAuthFailureFromSubscribeDbLock) {
...
+  rc = amf_handle_s6a_update_location_ans(&ula_ans);
+  EXPECT_EQ(rc, RETURNerror);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2604]: On failing to process, s6a-update-location -answer, I think AMF needs to reject the ongoing Registration procedure.

Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2604]: Actually we are updating location details earlier, because if there is any delay in updating location while starting the pdu session procedure, the pdu establishment request will be rejected, for avoiding this scenario we are updating location details before registration accepts.

If location update is failing then pdu session establishment should be rejected with dnn not matching.

Comment (Kaleem-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2604]: @rsarwad To add to what @Chandhu-Wavelabs have mentioned. 
s6a_update_location_ans_t is mainly used for validating the PDU session request w.r.t DNN configuration. This object is not used any where to validate registration request. So we are taking action of sending PDU session reject only when PDU session request is received i.e, if the PDU session request's DNN is not matched with the subscribed configuration from s6a_update_location_ans_t. Similarly we have taken action of sending PDU session reject if s6a response is NULL(Due to any exception from SubscriberDB).

Comment (rsarwad) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2604]: Is update location ans used only for DNN verification. Is this  aligned with spec? Because I went through the link, https://www.eventhelix.com/5G/standalone-access-registration/5g-standalone-access-registration.pdf and page number 5, which indicated that after AMF obtains subscriber details, AMF would send Registration Accept message in ICS.

Comment (Kaleem-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2604]: Yes you are correct, we are sending Registration Accept only after location information is received. 

In amf_handle_s6a_update_location_ans()
...
  /* FSM takes care of sending registration accept */
  ue_state_handle_message_initial(COMMON_PROCEDURE_INITIATED1,
                                  STATE_EVENT_SEC_MODE_COMPLETE, SESSION_NULL,
                                  ue_mm_context, amf_ctxt_p);
...

The reason why we are not sending registration reject when location information is received invalid/NULL. as we do not have any specific cause here to send (We are not referring this object during registration procedure). When we receive PDU session request and there if we do not have DNN value from location information answer we send PDU session reject with cause "DNN mismatch error"

Comment (rsarwad) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2604]: Thanks for explanation.

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 14]
```diff
@@ -2472,4 +2554,85 @@ TEST_F(AMFAppProcedureTest, TestAuthFailureFromSubscribeDbLock) {
...
+
+  ue_m5gmm_context_t* ue_context_p =
+      amf_ue_context_exists_amf_ue_ngap_id(ue_id);
+  // ue context should exist
+  ASSERT_NE(ue_context_p, nullptr);
+  memset(&ue_context_p->amf_context.apn_config_profile, 0,
+         sizeof(ue_context_p->amf_context.apn_config_profile));
+
+  /* Send uplink nas message for pdu session establishment request from UE */
+  rc = send_uplink_nas_pdu_session_establishment_request(
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2622]: Can we set expect for PDU session reject message?

Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2622]: // smf context should not be present
EXPECT_EQ(ue_context_p->amf_context.smf_ctxt_map.size(), 0);

We are checking here for smf context has been created or not, if pdu session establishment request is rejected then the smf_ctxt_map.size() should be zero.

Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:2622]: @rsarwad, we are defining event IDs to check expected events are happening or not.

std::vector<MessagesIds> expected_Ids{
      AMF_APP_NGAP_AMF_UE_ID_NOTIFICATION,  // new registration notification indication to ngap
      NGAP_NAS_DL_DATA_REQ,                 // Authentication Request to UE
      NGAP_NAS_DL_DATA_REQ,  // Security Command Mode Request to UE
      NGAP_NAS_DL_DATA_REQ,  // PDU Session Establishment Request with failure sm cause
      NGAP_NAS_DL_DATA_REQ,  // Deregistaration Accept
      NGAP_UE_CONTEXT_RELEASE_COMMAND  // UEContextReleaseCommand
  };
  
  we are checking the expected events and actual events are matching or not here
  EXPECT_EQ(expected_Ids, AMFClientServicer::getInstance().msgtype_stack);
  
After PDU session establishment request, we are checking smf context is present or not.
Case 1 : If context is present then pdu establishment is accepted and AMF sends downlink message to NGAP.
Case 2 : If context is not present then pdu establishment is rejected and AMF sends downlink message to NGAP.
Case 3 : If message dropped, there is no context and AMF does not send downlink message to AMF.

In this test, there is no smf context and AMF sending downlink to NGAP, means the downlink message is PDU session establishment reject.

cc : @ganeshg87 @panyogesh

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp]: addressed

**Review by rsarwad — APPROVED** [review]
LGTM

**Review by panyogesh — APPROVED** [review]

**Review by ssanadhya — APPROVED** [review]
